### DISCUSSION TITLE
Password reset link appears to be logging me back in, in both dev and hosted environments

### DISCUSSION THREAD

**Discussion by chrskerr** [supabase/supabase#1951] [Questions] [answered]
I have triggered the link by calling `supabase.auth.api.resetPasswordForEmail( username, { redirectTo: process.env.BASE_URL  });`

And when I click it, I end up logged back into the application. 

I have previously manually deleted all cookies, local storage and session storage to help ensure that no local state is being carried forward. 

If I reload the app without the query string from the link it remains logged out.

The only code that I believe is actively configuring supabase is onAuthStateChange, but that's mostly commented out right now.

Has anyone seen this happen before?

```
	useEffect(() => {
		const listener = supabase.auth.onAuthStateChange( async ( event, session ) => {
			console.log( "auth state changed", session, event );
			setSession( session );
			// if ( event === "PASSWORD_RECOVERY" ) setAuthView( "update_password" );
			// if ( event === "USER_UPDATED" ) setTimeout(() => setAuthView( "sign_in" ), 1000 );


		});
	
		return () => {
			listener.data.unsubscribe();
		};
	}, []);
```

**Comment by easy-street** [comment, marked as answer]
sounds similar to https://github.com/supabase/supabase/discussions/1920#discussioncomment-864182

the two lines commented out are important for properly handling password recovery links. :)

the `supabase` client detects & handles the URL query parameters and emits the `"PASSWORD_RECOVERY"` event. the commented lines set a local state value that indicates which view to render. when the `"PASSWORD_RECOVERY"` event is received, the app should render either the `@supabase/ui` `Auth.UpdatePassword` component, or a custom component with similar functionality.

the [nextjs-with-supabase-auth example](https://github.com/supabase/supabase/blob/master/examples/nextjs-with-supabase-auth/pages/index.js#L58) implements password recovery, so it's a helpful reference to step through in order to understand the flow.

**Reply by chrskerr** [reply]
Thanks @easy-street!

Is there a way to intercept the event and implement my own password reset form? 

Also do you know why the fallback behaviour is to behave like a magic link by any chance?

Cheers
chris

**Reply by easy-street** [reply]
> Is there a way to intercept the event and implement my own password reset form?

sure; you should be able to copy the `nextjs-with-supabase-auth` example implementation almost exactly, except for [this line](https://github.com/supabase/supabase/blob/master/examples/nextjs-with-supabase-auth/pages/index.js#L58):
```
{authView === 'forgotten_password' && <Auth.UpdatePassword supabaseClient={supabase} />}
```

instead of rendering `Auth.UpdatePassword` when `authView` is `forgotten_password`, render a custom password reset form component. for example:

```
{authView === 'forgotten_password' && <UpdatePassword  />}
```

minimal `UpdatePassword` implementation:
```
function UpdatePassword() {
  const inputRef = React.useRef();

  async function handleSubmit(event) {
    event.preventDefault();
    await supabase.auth.update({ password: inputRef.current.value });
  }

  return (
    <form onSubmit={handleSubmit}>
      <input ref={inputRef} type="password" />
      <button type="submit">Update Password</button>
    </form>
  );
}
```

> Also do you know why the fallback behaviour is to behave like a magic link by any chance?

The supabase implementation to send & handle password recovery links essentially has two goals:
1) sign in the user (since `supabase.auth.update` will update the current session's user)
2) emit the `"PASSWORD_RECOVERY"` event so the app can respond to the event by rendering the `Auth.UpdatePassword` component from `@supabase/ui`, or a custom component.

if the event is ignored, the outcome is effectively the same as a magic link (user is signed in).

**Reply by chrskerr** [reply]
Thanks heaps for all of the help, I now understand. I was stuck in the Firebase mental model and was confused, but this approach makes lots of sense!

Cheers!
